HW_2.2/hw_2.py

```diff
@@ -0,0 +1,342 @@
+    + Нужно написать методы `read_record` и `read_records`, которые по смыслу соответствуют `readline()` и `readlines()` в обычных файлах, но они должны выдавать не строки, а объект(ы) `FastaRecord`
+2. Конструктор должен принимать один аргумент - **путь к файлу**
+3. Класс должен эффективно распоряжаться памятью, с расчётом на работу с очень большими файлами
+    
+Объект `FastaRecord`. Это должен быть **датакласс** (см. про примеры декораторов в соответствующей лекции) с тремя полями:
++ `seq` - последовательность
++ `id_` - ID последовательности (это то, что в фаста файле в строке, которая начинается с `>` до первого пробела. Например, >**GTD326487.1** Species anonymous 24 chromosome) 
++ `description` - то, что осталось после ID (Например, >GTD326487.1 **Species anonymous 24 chromosome**)
+
+
+Напишите демонстрацию работы кода с использованием всех написанных методов, обязательно добавьте файл с тестовыми данными в репозиторий (не обязательно большой)
+
+**Можно использовать модули из стандартной библиотеки**
+"""
+
+# Ваш код здесь
+import os
+from dataclasses import dataclass
+
+@dataclass
+class FastaRecord:
+    seq: str
+    id_: str
+    description: str
+
+class OpenFasta:
+    def __init__(self, path):
+        self.path = path
+        self.__file = None
+        
+    def __enter__(self):
+        self.open()
+        return self
+    
+    def __exit__(self, exc_type, exc_val, exc_tb):
+        self.__file = None
+        return False
+        
+    def open(self):
+        self.__file = open(self.path, 'r')
+        
+    def close(self):
+        if self.__file is not None:
+            self.__file.close()
+            self.__file = None
+            
+    def read_record(self):
+        if self.__file is None:
+            raise IOError('Fasta file was not opened')
+        id = None
+        description = None
+        seq = ""
+        while True:
+            pos = self.__file.tell()
+            first_char = self.__file.read(1)
+            self.__file.seek(pos)
+            if first_char == '':
+                if id is not None:
+                    return FastaRecord(seq, id, description)
+                else:
+                    return None
+            if first_char == '>':
+                if id is not None:
+                    return FastaRecord(seq, id, description)
+            line = self.__file.readline()
+            if line.startswith('>'):
+                id, description = line[1::].split(' ', 1)
+                description = description[:-1]
+            else:
+                seq += line[:-1]
+            
+    def read_records(self):
+        while True:
+            record = self.read_record()
+            if record is None:
+                break
+            yield record
```

Comment:
readlines возвращает не генератор, а список - именно поэтому его и не рекомендуют использовать на больших файлах.
Как вариант - что-то такое можно было сделать.
```suggestion
    def read_records(self):
        records_list = []
        while True:
            try:
                records_list.append(self.read_record())
            except StopIteration:
                return records_list
        return records_list
```